src/SignalR/common/Shared/AsyncEnumerableAdapters.cs

```diff
@@ -13,38 +13,63 @@ namespace Microsoft.AspNetCore.SignalR.Internal
+            var enumerator = asyncEnumerable.GetAsyncEnumerator(cancellationToken);
+            return enumerator as IAsyncEnumerator<object?> ?? new BoxedAsyncEnumerator<T>(enumerator);
         }
 
         public static IAsyncEnumerable<T> MakeCancelableTypedAsyncEnumerable<T>(IAsyncEnumerable<T> asyncEnumerable, CancellationTokenSource cts)
         {
             return new CancelableTypedAsyncEnumerable<T>(asyncEnumerable, cts);
         }
 
-#if NETCOREAPP
-        public static async IAsyncEnumerable<object?> MakeAsyncEnumerableFromChannel<T>(ChannelReader<T> channel, [EnumeratorCancellation] CancellationToken cancellationToken = default)
+        public static IAsyncEnumerator<object?> MakeAsyncEnumeratorFromChannel<T>(ChannelReader<T> channel, CancellationToken cancellationToken = default)
         {
-            await foreach (var item in channel.ReadAllAsync(cancellationToken))
-            {
-                yield return item;
-            }
+            return new ChannelAsyncEnumerator<T>(channel, cancellationToken);
         }
-#else
-        // System.Threading.Channels.ReadAllAsync() is not available on netstandard2.0 and netstandard2.1
-        // But this is the exact same code that it uses
-        public static async IAsyncEnumerable<object?> MakeAsyncEnumerableFromChannel<T>(ChannelReader<T> channel, [EnumeratorCancellation] CancellationToken cancellationToken = default)
+
+        private class ChannelAsyncEnumerator<T> : IAsyncEnumerator<object?>
         {
-            while (await channel.WaitToReadAsync(cancellationToken).ConfigureAwait(false))
+            private readonly ChannelReader<T> _channel;
+            private readonly CancellationToken _cancellationToken;
+            public object? Current { get; private set; }
+
+            public ChannelAsyncEnumerator(ChannelReader<T> channel, CancellationToken cancellationToken)
+            {
+                _channel = channel;
+                _cancellationToken = cancellationToken;
+            }
+
+            public ValueTask<bool> MoveNextAsync()
+            {
+                var result = _channel.ReadAsync(_cancellationToken);
+
+                if (result.IsCompletedSuccessfully)
+                {
+                    Current = result.Result;
+                    return new ValueTask<bool>(true);
+                }
+
+                return new ValueTask<bool>(MoveNextAsyncAwaited(result));
+            }
+
+            private async Task<bool> MoveNextAsyncAwaited(ValueTask<T> channelReadTask)
             {
-                while (channel.TryRead(out var item))
+                try
+                {
+                    Current = await channelReadTask.ConfigureAwait(false);
+                }
+                catch (ChannelClosedException ex) when (ex.InnerException == null)
                 {
-                    yield return item;
+                    return false;
```

Comment:
I think it would be better to call `Channel.TryRead()` in `MoveNextAsync()` and call `Channel.WaitToReadAsync()` (and then `Channel.TryRead()` again if necessary) here so we don't always need to observe the completion of this Channel with a `ChannelClosedException`.

Reply:
ack